### DISCUSSION TITLE
How to properly generate diagnostics ? (no lint warnings emitted by slang)

### DISCUSSION THREAD

**Discussion by suzizecat** [MikePopoloski/slang#908] [Q&A]
Hi ! 

In my effort to make a language server using Slang, I am having an issue to get warnings which are not related to the syntax such as 

```
 warning: implicit conversion truncates from 32 to 2 bits [-Wwidth-trunc]
```

I, however, do get any syntax error, duplicated files problems and such as expected but cannot manage to have "lint" warnings to be emitted by slang.

Would you have some pointer in troubleshooting this problem ?

---

I use the following code, called each time the workspace is re-evaluated to check for new problems (where `slsp` is my project namespace).

There is no expected diagnostic filtering on the diagnostic client.

For reference, the code is [this one](https://github.com/suzizecat/slang-lsp-tools/blob/073877eb5e9a51a2fb0ce35b9e6ce03ece4eeec9/lsp-server/diplomat/src/diplomat_lsp.cpp#L223), more or less some minor details. 

```cpp
    // As per slang limitation, it is needed to recreate the diagnostic engine
    // because the source manager will be deleted by _read_workspace_modules.
    // Therefore, the client shall also be rebuilt.
    _read_workspace_modules(); // Regenerate syntax trees for each workspace files
    _diagnostic_client.reset(new slsp::LSPDiagnosticClient(_documents,_sm.get(),_diagnostic_client.get()));

    slang::DiagnosticEngine de = slang::DiagnosticEngine(*_sm);

    de.setErrorLimit(500);
    de.setIgnoreAllWarnings(false);
    de.setSeverity(slang::diag::MismatchedTimeScales,slang::DiagnosticSeverity::Ignored);
    de.setSeverity(slang::diag::MissingTimeScale,slang::DiagnosticSeverity::Ignored);
    de.addClient(_diagnostic_client);


    slang::ast::CompilationOptions coptions;
    coptions.flags &= ~ (slang::ast::CompilationFlags::SuppressUnused);
    slang::Bag bag(coptions);
    _compilation.reset(new slang::ast::Compilation(bag)); //_compilation is a smart pointer

    spdlog::info("Add syntax trees");
    for (const auto& [key, value] : _documents) // More or less a collection of syntax trees
    {
        _compilation->addSyntaxTree(value->st);
    }

    _compilation->getRoot();

    spdlog::info("Issuing diagnostics");
    for (const slang::Diagnostic& diag : _compilation->getAllDiagnostics())
        de.issue(diag);

    spdlog::info("Send diagnostics");
    _emit_diagnostics(); // Actually send all diagnostic through the LSP

```

**Comment by MikePopoloski** [comment]
Hmm, your code looks correct to me. A guess would be that the warnings you want to see are in uninstantiated modules? The set of diagnostics slang will issue for those modules is limited because they don't always make sense when not part of the elaborated design tree. Do you have an example source file that you want to see warnings for?

**Reply by suzizecat** [reply]
I have not specified a top module, so maybe a wrong one has been selected and so warnings are not issued.
I will investigate with a top-level (or at least a fileset that has only one top level).

**Reply by suzizecat** [reply]
Using current master with a reasonable fileset, the warnings actually shows up.
I will try again on my initial testcase.

**Reply by suzizecat** [reply]
So, using the exact same code as yesterday but with a bigger codebase, the warnings disappears.
Unfortunately I cannot share this codebase and I am not really able to pinpoint the issue.

For reference, my workspace includes quite a bit of "generic" material that may or may not be used by several numbers of projects (you may see it as a toolbox of SV modules). Therefore my workspace has a lot of unused module in the filetree that I would rather not filter out (for simplicity). 
The full file tree contains 80 files, of which only one is actually my design (for the sake of testing, through explicitly setting the top level).

After experimenting, it looks like removing a certain amount of files from my fileset allows the warnings to be generated.
The threshold seems to be 70 files (70 files : no diagnostic, 69 files, got diagnostics), most of which are *not* used by the top level.

Moreover, the actually removed files have no impact (as long as they are not used by the selected design, of course) 

Ignoring `diag::UnusedDefinition` does not have an impact either.

Does this threshold rings any bell to you ? 

## Tests

With slang 5.0, using 
```shell
$ slang -f test_filelist.lst --top my_top -Weverything --error-limit=500
```
with  my full fileset in the file list works as expected and emits 79 errors, 61 warnings (allof errors being timescale issues) that disapear when using 

```shell
$ slang -f test_filelist.lst --top my_top -Weverything --error-limit=500 --timescale=1ns/1ps
```
Yielding  0 errors, 63 warnings.
